plugins/wporg-5ftf/includes/stats.php

```diff
@@ -0,0 +1,185 @@
+<?php
+
+/**
+ * Track and display key metrics for the program, to measure growth and effectiveness.
+ */
+
+namespace WordPressDotOrg\FiveForTheFuture\Stats;
+
+use WordPressDotOrg\FiveForTheFuture;
+use WordPressDotOrg\FiveForTheFuture\{ Contributor, Pledge, XProfile };
+use WP_Query;
+
+defined( 'WPINC' ) || die();
+
+add_action( 'init',                 __NAMESPACE__ . '\register_post_types' );
+add_action( 'init',                 __NAMESPACE__ . '\schedule_cron_jobs' );
+add_action( '5ftf_record_snapshot', __NAMESPACE__ . '\record_snapshot' );
+
+add_shortcode( '5ftf_stats', __NAMESPACE__ . '\render_shortcode' );
+
+
+/**
+ * Register the snapshots post type. Each post represents a snapshot of the stats at a point in time.
+ */
+function register_post_types() {
+	$args = array(
+		'supports'     => array( 'custom-fields' ),
+		'public'       => false,
+		'show_in_rest' => true,
+
+		// Only allow posts to be created programmatically.
+		'capability_type' => '5ftf_stats',
+		'capabilities'    => array(
+			'create_posts' => 'do_not_allow',
+		),
+	);
+
+	register_post_type( '5ftf_stats_snapshot', $args );
+}
+
+/**
+ * Schedule the cron job to record a stats snapshot.
+ */
+function schedule_cron_jobs() {
+	if ( wp_next_scheduled( '5ftf_record_snapshot' ) ) {
+		return;
+	}
+
+	// Schedule a repeating "single" event to avoid having to create a custom schedule.
+	wp_schedule_single_event(
+		time() + ( 2 * WEEK_IN_SECONDS ),
+		'5ftf_record_snapshot'
+	);
+}
+
+/**
+ * Record a snapshot of the current stats, so we can track trends over time.
+ */
+function record_snapshot() {
+	$stats = get_snapshot_data();
+
+	$post_id = wp_insert_post( array(
+		'post_type'   => '5ftf_stats_snapshot',
+		'post_author' => 0,
+		'post_title'  => sprintf( '5ftF Stats Snapshot %s', date( 'Y-m-d' ) ),
+		'post_status' => 'publish',
+	) );
+
+	add_post_meta( $post_id, '5ftf_total_pledged_hours',             $stats['confirmed_hours'] );
```

Comment:
Do you think this should use `PledgeMeta\META_PREFIX`, or create a new stats prefix constant, rather than hardcoding this? (I don't really have a preference, tbh, but we do that in PledgeMeta)

Reply:
I think it would be good to stay consistent with the rest of the plugin and at least use `FiveForTheFuture\PREFIX` instead of hardcoding `5ftf`.

Reply:
I'll add it here for consistency, but in general I think using constants in cases like this makes the code harder to work with.

### Pros:

* All usages of it will be consistent; e.g., no human error where hardcoding ends up mixing `5ftf_` and `5ftf-`

* If we ever want to rename it, we only have to change it in 1 place. 

    Changing a prefix like that would be very rare, and probably create a bunch of related headaches (having to rename database content, etc), so I think it's very unlikely.

Do either of you feel like there are any that I missed?

### Cons:

* It's less readable,  and more abstract / less obvious: With `5ftf_foo_bar` I instantly have everything I need to know; I just continue reading without even thinking about it. With `PREFIX . 'foo_bar'` it takes extra mental cycles to translate the abstraction into the actual value.

    If I haven't memorized the exact value of `PREFIX` in this file, then sometimes I have to have stop and wonder if `PREFIX` is `5ftF_` or `5ftf-` or  `5ftf_stats_` or `five_` or any of the other possible combinations. 

    When using the prefix, I sometimes have to try and remember if it includes the `_` or not, since `PledgeMeta\META_PREFIX` does and `FiveForTheFuture\PREFIX` doesn't. 

    In all of that, I'm pulled out of my train of thought, and have to context-switch to jump to the definition get find the exact value, then jump back to where I was and pick back up where I left off.   

* If I'm in the DOM inspector, a database browser, etc and I see an HTML field/nonce/whatever like `5ftf_stats`, and I want to find out where that's generated in the source, I can't just grep for `5ftf_stats`. In order to know if I need to strip off the prefix when searching, I have to stop and try to remember if the code is using a prefix for that particular thing or not.

   If I _do_ have to strip the prefix off, that often results in a ton of false positives because the query is just `stats`. In many cases I can narrow it to `'stats'` by guessing how the name was constructed, but there are still times where that doesn't work because `"` was used instead of `'`, etc. 

    So I often end up having to run the search multiple times; first with the prefix to hopefully get results with a low signal/noise ratio, and then run several other variations if it turns out it was using a prefix. In the latter case I then have to dig through dozens/hundreds of results to find the one I'm looking for.

Abstractions always have a cost associated with them, and in this case I think that cost is higher than the benefits it provides.